### PR TITLE
added instancesdk.List() feature and its gRPC API

### PR DESCRIPTION
**Pull request by ilgooz**
depends on https://github.com/mesg-foundation/engine/pull/1074

### CODE REVIEW

File: protobuf/api/instance.proto [hunk 1]
```diff
@@ -1,12 +1,23 @@
 syntax = "proto3";
 
+import "protobuf/definition/instance.proto";
+
 package api;
 
 service Instance {
+  rpc List (ListInstancesRequest) returns (ListInstancesResponse) {}
```
Comment (NicolasMahe) [inline protobuf/api/instance.proto]: Same comment as https://github.com/mesg-foundation/engine/pull/1075/files#r295623883

File: protobuf/api/instance.proto [hunk 2]
```diff
@@ -1,12 +1,23 @@
...
   rpc Create (CreateInstanceRequest) returns (CreateInstanceResponse) {}
   rpc Delete (DeleteInstanceRequest) returns (DeleteInstanceResponse) {}
 }
 
+message ListInstancesRequest {
+  string serviceHash = 1;   // service hash of instance to filter by service (optional).
```
Comment (NicolasMahe) [inline protobuf/api/instance.proto]: This filter is actually not optional.
There is no test on its default value in `sdk/instance/instance.go` or `database/instance_db.go`

Comment (ilgooz) [reply protobuf/api/instance.proto]: do you want to make it optional or rm the optional word from proto?

Comment (NicolasMahe) [reply protobuf/api/instance.proto]: make it optional ;)

Comment (ilgooz) [reply protobuf/api/instance.proto]: done

File: database/instance_db.go [hunk 3]
```diff
@@ -12,6 +12,12 @@ type InstanceDB interface {
 	// Get retrives instance by instance hash.
 	Get(hash string) (*instance.Instance, error)
 
+	// GetAll retrives all instances.
```
Comment (NicolasMahe) [inline database/instance_db.go]:
```suggestion
	// GetAll retrieves all instances.
```

File: database/instance_db.go [hunk 4]
```diff
@@ -12,6 +12,12 @@ type InstanceDB interface {
...
+	GetAll() ([]*instance.Instance, error)
+
+	// GetAllByService retrives all instances of service by service's hash.
```
Comment (NicolasMahe) [inline database/instance_db.go]:
```suggestion
	// GetAllByService retrieves all instances of service by service's hash.
```

File: database/instance_db.go [hunk 5]
```diff
@@ -63,6 +69,38 @@ func (d *LevelDBInstanceDB) Get(hash string) (*instance.Instance, error) {
 	return d.unmarshal(hash, b)
 }
 
+// GetAll retrives all instances.
```
Comment (NicolasMahe) [inline database/instance_db.go]:
```suggestion
// GetAll retrieves all instances.
```

File: database/instance_db.go [hunk 6]
```diff
@@ -63,6 +69,38 @@ func (d *LevelDBInstanceDB) Get(hash string) (*instance.Instance, error) {
...
+			iter.Release()
+			return nil, err
+		}
+		instances = append(instances, i)
+	}
+	iter.Release()
+	return instances, iter.Error()
+}
+
+// GetAllByService retrives all instances of service by service's hash.
```
Comment (NicolasMahe) [inline database/instance_db.go]:
```suggestion
// GetAllByService retrieves all instances of service by service's hash.
```

**Pull request merged**